Hlu/summarization evaluation

hlums opened the pull request:
### Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->


### Related Issues
<!--- If it fixes an open issue, please link to the issue here. -->


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [x] I have added tests.
- [x] I have updated the documentation accordingly.

daden-ms commented on examples/text_summarization/summarization_evaluation.ipynb:
```diff
@@ -0,0 +1,391 @@
+{
```
<p>adjcent&nbsp;typo</p><p>sentence level&nbsp;-&gt; sentence-level&nbsp;</p><p>Would it better to use <a href="https://github.com/bheinzerling/pyrouge" target="_blank">https://github.com/bheinzerling/pyrouge</a> instead of <a href="https://github.com/bheinzerling/pyrouge/tree/master/pyrouge" target="_blank">https://github.com/bheinzerling/pyrouge/tree/master/pyrouge</a>?</p><p>other languages-&gt; other natural languages?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp-recipes/pull/526/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

hlums replied:
Fixed 1, 2, and 3. For 4, I think it's clear from the context that I don't mean programming languages.

daden-ms commented on examples/text_summarization/summarization_evaluation.ipynb:
```diff
@@ -0,0 +1,391 @@
+{
```
<p>This is really nice. I do remember some setup is required for using cpan.  Not sure if that can be automated. Or maybe at least mention it. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp-recipes/pull/526/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

hlums replied:
<p>Do you remember what else needs to be done to use cpan? I've tested these steps on two machines. </p>

daden-ms replied:
just config the cpan.

hlums replied:
Any specific commands? I don't remember anything else besides these commands. If you don't remember, we can add it later if people run into problems using these commands.

daden-ms commented on examples/text_summarization/summarization_evaluation.ipynb:
```diff
@@ -0,0 +1,391 @@
+{
```
<p>Does compute_rouge_perl support other natural languages? If so, can you provide a similar Hindi example; if not, can you mention it in the writing?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp-recipes/pull/526/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

hlums replied:
No, compute_rouge_perl doesn't support other languages. It's going to be hard to hack it to do it since I don't know perl. I mentioned it in the introduction section.

daden-ms commented on tests/unit/test_eval_compute_rouge.py:
```diff
@@ -0,0 +1,221 @@
+    pytest.approx(rouge_perl["rouge_l_f_score"], RLF, abs=ABS_TOL)
+
+
+def test_compute_rouge_python_file(rouge_test_data, tmp):
```
use tmp fixture to make sure the tmp folder clean up after testing

hlums replied:
I'm using the tmp fixure here.

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+from .rouge_ext import RougeExt
+
+
+def compute_rouge_perl(cand, ref, input_files=False):
```
use is_input_files as a better name

hlums replied:
How about input_is_files?

daden-ms replied:
in python, i think the convention is to use is*, like isnumeric().

hlums replied:
OK. That makes sense. Changed it to is_input_files

hlums replied:
OK. That makes sense. Changed it to is_input_files

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+            with open(tmp_dir + "/reference/ref.{}.txt".format(i), "w", encoding="utf-8") as f:
+                f.write(references[i])
+        r = Rouge155()
+        r.model_dir = tmp_dir + "/reference/"
```
move this ahead before line 51 and 52 to avoid repeating the code.

hlums replied:
Fixed

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+        for i in range(cnt):
+            if len(references[i]) < 1:
+                continue
+            with open(tmp_dir + "/candidate/cand.{}.txt".format(i), "w", encoding="utf-8") as f:
```
same here. the path can be saved to avoid having the same constant multiple times in the code.

hlums replied:
Fixed

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+        r.model_filename_pattern = "ref.#ID#.txt"
+        r.system_filename_pattern = r"cand.(\d+).txt"
+        rouge_results = r.convert_and_evaluate()
+        print(rouge_results)
```
result seems very long, we can have a verbose option to have these printouts only when verbose is True.

hlums replied:
Done.

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+        print(rouge_results)
+        results_dict = r.output_to_dict(rouge_results)
+    finally:
+        pass
```
pass should be removed

hlums replied:
Thanks for catching it! Fixed.

daden-ms commented on utils_nlp/eval/compute_rouge.py:
```diff
@@ -0,0 +1,127 @@
+    return results_dict
+
+
+def compute_rouge_python(cand, ref, input_files=False, language="en"):
```
consider use is_input_files

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+# This script is adopted from https://github.com/Diego999/py-rouge/blob/master/rouge/rouge.py
```
copyright

daden-ms replied:
can all the python files related to rouge move to a folder,  eval/rouge?

hlums replied:
According to CELA, we don't need to add the MSFT copyright if we don't make significant changes to a piece of code, which is the case here.
I created a rouge folder and moved rouge related scripts there.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+        else:
+            self.stemmer = RougeExt.STEMMER_DICT[self.language]
+
+        # # Load static objects
```
can we just remove these commented-out code lines?

hlums replied:
I prefer keeping these to keep record of what I changed to the original code. I found this useful when I looked at how BERTSum changed the rouge scripts.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+            # sentence = Rouge.REMOVE_CHAR_PATTERN.sub(' ', sentence.lower()).strip()
+            sentence = self.remove_char_pattern.sub(" ", sentence.lower()).strip()
+
+            #     # Preprocess. Hack: because official ROUGE script bring "cannot" as "cannot"
```
remove these commented lines if they are not relevant comments.

hlums replied:
I prefer keeping these to keep record of what I changed to the original code.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+
+        self.language = language
+
+        self.sentence_split = RougeExt.SENTENCE_SPLIT_DICT[self.language]
```
can we make sentence_split, word_tokenize, remove_char_pattern and stemmer as arguments so that this class can be used for other languages without specifically editing the dictionary inside this class?

hlums replied:
Agree that makes the code more flexible, but makes the function harder to use for languages we already support. I added the arguments so that the user can provide these if the language is not already supported by the internal dictionary. Please take a look.

daden-ms replied:
By "the function harder to use for languages we already support", which function do you mean?  I didn't see the arguments for these functions in the __init__ function.

hlums replied:
Sorry, I actually meant the RougeExt class, but also the compute_rouge_python function. I added the additional arguments to both. You need to go to the latest commit to see the changes. 
https://github.com/microsoft/nlp-recipes/blob/d4c2a26536a6e0ad5e28a6a34a3736d985b054e3/utils_nlp/eval/rouge/compute_rouge.py#L86
https://github.com/microsoft/nlp-recipes/blob/d4c2a26536a6e0ad5e28a6a34a3736d985b054e3/utils_nlp/eval/rouge/rouge_ext.py#L67

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+        Returns:
+          List of tokens of text
+        """
+        return self.word_tokenize(text, self.language)
```
is single line function needed?

hlums replied:
Good point! This function was in the original code and I prefer keeping it to reduce changes to the original code.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+          List of tokens of text
+        """
+
+        return self.sentence_split(text, self.language)
```
is single line function needed?

hlums replied:
Good point! This function was in the original code and I prefer keeping it to reduce changes to the original code.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+# 2) word tokenizer (WORD_TOKENIZE_DICT)
+# 3) pattern of characters to remove (REMOVE_CHAR_PATTERN_DICT)
+# 4) stemmer (STEMMER_DICT), this is optional since stemming is not applicable to all languages
+# 5) changes to _split_into_words if the text words are not separated by space
```
_split_into_words still splits words by space, in the current version.

hlums replied:
Yes. I meant if you want to support languages like Chinese, you need to change this function to handle text that's not separated by space.

daden-ms commented on utils_nlp/eval/rouge_ext.py:
```diff
@@ -0,0 +1,978 @@
+from ..language_utils.hi.hindi_stemmer import hi_stem
+
+
+class RougeExt:
```
can you write RougeExt in a way to inherit Rouge, so that the functions that you didn't change don't need to repeat themselves here?

hlums replied:
That's a good point! I can try that. It will make the code look much cleaner. 
However, I'm a little concerned about static methods not needed by this class like load_stemmer and load_wordnet_db getting inherited and causing confusion. Thoughts?

daden-ms replied:
I think the purpose here is to really capture the change you made and make the knowledge easier to understand. It might involve import the original Rouge, making simpler changes to it and then write an extensible version to it.

hlums replied:
Yes. I think it's a good idea to inherit from Rouge and did it. 
Then I had to change some static methods to non-static because of the custom options added. For example, to add a custom word splitter, I add to change _split_into_words to non-static, then _get_word_ngrams_and_length and _get_unigrams, then _compute_ngrams and _compute_ngrams_lcs. So the code isn't as clean as I was expecting. 
Please take a look and let me know if you have any suggestions to further simplify it.

hlums replied:
Yes. I think it's a good idea to inherit from Rouge and did it.
Then I had to change some static methods to non-static because of the custom options added. For example, to add a custom word splitter, I add to change _split_into_words to non-static, then _get_word_ngrams_and_length and _get_unigrams, then _compute_ngrams and _compute_ngrams_lcs. So the code isn't as clean as I was expecting.
Please take a look and let me know if you have any suggestions to further simplify it.

daden-ms commented on utils_nlp/language_utils/hi/hindi_stemmer.py:
```diff
@@ -0,0 +1,101 @@
+#! /usr/bin/env python3.1
```
please cite the source of this file.

hlums replied:
I kept the original script as it is, which already contains copyright and citation info.

daden-ms replied:
I meant to ask if you downloaded the code from https://research.variancia.com/hindi_stemmer/ and if so, link this page.

hlums replied:
Added the link.

hlums commented:
@daden-ms I've addressed all comments so far. Let me know if you have any additional comments on the changes.

daden-ms commented:
> @daden-ms I've addressed all comments so far. Let me know if you have any additional comments on the changes.

There is discrepancy between what you handle the rouge_ext and compute_rouge_python. Let's discuss the change for Rouge_ext offline.

daden-ms commented on utils_nlp/eval/rouge/compute_rouge.py:
```diff
@@ -0,0 +1,161 @@
+from .rouge_ext import RougeExt
+
+
+def compute_rouge_perl(cand, ref, is_input_files=False, verbose=False):
```
for this one, can you add sentence splitting token? I believe this is the common customization that's needed.

hlums replied:
Can you point me to a script that uses the "sentence splitting token"? I didn't find it in Rouge155 or the BERTSum test_rouge function. I probably missed it.

daden-ms replied:
here it splits the sentences by using "<q>": https://github.com/daden-ms/BertSum/blob/master/bertsum/others/pyrouge.py#L254

hlums replied:
I see. Since this option is not exposed to users by Rouge155, we won't be able to control it without changing the underlying pyrouge package. I prefer not changing the underlying pyrouge package if possible. It's better to take dependency on a standard package instead of creating our own variation for a small change like this.
I think BERTSum did it specifically for CNN/DailyMail. Is it possible to preprocess CNN/DM to use <\n> as line separator instead of <q>?

daden-ms replied:
The harvardnlp CNN/DM uses <q> in the summary but we can replace this with a new line when calculating rouge score. The result is close enought, as indicated by https://github.com/google-research/google-research/issues/168

hlums commented:
> > @daden-ms I've addressed all comments so far. Let me know if you have any additional comments on the changes.
> 
> There is discrepancy between what you handle the rouge_ext and compute_rouge_python. Let's discuss the change for Rouge_ext offline.

Made the changes we discussed. Let me know if missed anything.
